functionary/ui/templatetags/task_output.py

```diff
@@ -0,0 +1,134 @@
+    """
+    if result_type == "string":
+        data = io.StringIO(result)
+        csv_data = csv.reader(data)
+        result = {"headers": next(csv_data), "data": [row for row in csv_data]}
+    elif result_type == "json":
+        json_data = result
+        if isinstance(result, str):
+            json_data = json.loads(result)
+
+        if len(json_data) > 0:
+            data = []
+            if isinstance(json_data, list):
+                data = json_data
+            elif isinstance(json_data, dict) and len(json_data.values()) == 1:
+                values = next(iter((json_data.values())))
+                if isinstance(values, list):
+                    data = values
+
+            if len(data) > 0:
+                headers = data[0].keys()
+                res_data = []
+                for row in data:
+                    r = []
+                    for key in headers:
+                        r.append(row[key])
+                    res_data.append(r)
+
+                result = {"headers": headers, "data": res_data}
+            else:
+                raise ValueError("Unable to determine the structure of the JSON")
+        else:
+            raise ValueError("JSON data appears to be empty")
+    else:
+        raise ValueError(f"Unable to convert {result_type} to csv")
+    return result
+
+
+# Add table first since we can convert some JSON to CSV
+format_mapper = {
+    "table": _format_table,
+    "json": _format_json,
+}
+
+
+class TaskOutputNode(Node):
+    def __init__(self, task):
+        self.task = Variable(task)
+
+    def render(self, context):
+        the_task = self.task.resolve(context)
+        if not the_task.result:
+            return "<span>No results</span>"
+
+        result = the_task.result
+        return_type = the_task.return_type
+        output_format = the_task.output_format
+
+        formats_to_try = [output_format] if output_format else format_mapper.keys()
+
+        # If the return type isn't set, do our best guess to
+        # generate a useful output
+        if not return_type and isinstance(result, str):
+            return_type = "string"
+
+        for fmt in formats_to_try:
+            try:
+                ctx = {
+                    "format": fmt,
+                    "result": format_mapper[fmt](result, return_type),
+                }
+
+                plate = get_template(f"tags/output_{fmt}.html")
+                return plate.render(context=ctx)
+            except Exception as e:
+                logger.debug(
+                    "Unable to convert output for %s to %s: %s",
+                    the_task.id,
+                    fmt,
+                    e,
+                )
+
+        return f"<pre>{'<br/>'.join(the_task.result.splitlines())}</pre>"
+
+
+@register.tag("task_output")
+def do_task_output(parser, token):
+    """Convert the result of the task to the specified format."""
+    _, rest = token.contents.split(None, 1)
```

Comment:
I don't know what the right name for `rest`, but it's not obvious what data that refers to.  Is `tag_param` or something similar a better name?

Reply:
yes, that was a bad name